### PR TITLE
feat: added support for message timestamp

### LINKED ISSUE #912: get message publish timestamp
Hi

I am looking to get message publish timestamp on the returned json payload from consumer subscription.
Any option or ways to get this info?

### PR DESCRIPTION
**Pull request by antonio-pedro99**
**Description**:
This PR fixes #912. It introduces enhancements to the Kafka bridge to support more flexible message timestamp handling by allowing producers to specify timestamps explicitly in `ProducerRecord` objects. Currently, the Kafka bridge does not fully utilize the `timestamp` parameter in the `ProducerRecord` constructors, leading to using the CreateTime message timestamp type every time we produce a message. On the other hand, the `ConsumerRecord` has also been updated with the timestamp parameter.

**Changes**:
- Implemented support for interpreting the `timestamp` parameter in `ProducerRecord` objects sent to Kafka topics via the bridge.
- Allow users to read the `ConsumerRecord` timestamp on the request's response. E.g

```json
[
    {
        "topic": "bridge-quickstart-topic",
        "key": "my-key",
        "value": "sales-lead-0001",
        "partition": 0,
        "offset": 0,
        "timestamp": 1719663008510(new)
    },
    {
        "topic": "bridge-quickstart-topic",
        "key": "my-key",
        "value": "sales-lead-0001",
        "partition": 0,
        "offset": 1,
        "timestamp": 1719663054984(new)
    }
]
```
- Updated the Kafka bridge logic to correctly handle timestamps specified by producers, aligning with configured Kafka broker timestamp policies (`CreateTime` or `LogAppendTime`).

**Testing**:
- Unit tests were added to verify the correct interpretation and propagation of `timestamp` in `ProducerRecord`.
- I am still adjusting to the `BasicKafkaClient` to change `ConsumerIT` with the required tests.

### CODE REVIEW

**Comment by ppatierno** [comment]
@antonio-pedro99 I haven't reviewed the PR yes but I see this doc build failure here you should address ...

```shell
ERROR: Uncommitted changes in documentation:
M	documentation/book/api/definitions.adoc
Run the following to add up-to-date resources:
  make docu_api \
    && git add documentation/book/ \
    && git commit -s -m 'Update generated documentation'
make: *** [Makefile:67: docu_check] Error 1
```

after that I will start my review. Thanks!

**Comment by antonio-pedro99** [comment]
> ```shell
> make docu_api \
>     && git add documentation/book/ \
>     && git commit -s -m 'Update generated documentation'
> ```

Any clue about this error?

```bash
[ERROR] Failed to execute goal io.github.swagger2markup:swagger2markup-maven-plugin:1.3.7:convertSwagger2markup (generate-apidoc) on project kafka-bridge: Failed to execute goal 'convertSwagger2markup': Error creating extended parser class: Could not determine whether class 'org.pegdown.Parser$$parboiled' has already been loaded: Unable to make protected final java.lang.Class java.lang.ClassLoader.findLoadedClass(java.lang.String) accessible: module java.base does not "opens java.lang" to unnamed module @358ab600 -> [Help 1]
```

I have been trying to fix, but could figure it out.

**Comment by ppatierno** [comment]
> Any clue about this error?

No idea :-(

**Comment by scholzj** [comment]
Could it be a wrong Java version or something?

**Comment by ppatierno** [comment]
@antonio-pedro99 we are going to plan a release. Are you still working to fix this on your side? or should we leave it out of 0.30.0?

**Comment by antonio-pedro99** [comment]
> @antonio-pedro99 we are going to plan a release. Are you still working to fix this on your side? or should we leave it out of 0.30.0?

Yeah, I am going to finish it soon.

**Comment by antonio-pedro99** [comment]
> Could it be a wrong Java version or something?

I guess so

**Comment by antonio-pedro99** [comment]
@ppatierno, we had a similar issue here https://github.com/strimzi/strimzi-kafka-bridge/pull/858#issuecomment-1872924723
Maybe we should try to push the files again?

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 1]
```diff
@@ -57,11 +60,11 @@ public int sendStringMessagesPlain(long timeoutMs, String topicName, int message
...
-        try (Producer plainProducer = new Producer.ProducerBuilder(resultPromise, msgCntPredicate, topicName, message, partition)
-            .withProperties(properties)
-            .withHeaders(headers)
-            .withNullKeyRecord(withNullKeyRecord)
-            .build()) {
+        try (Producer plainProducer = new Producer.ProducerBuilder(resultPromise, msgCntPredicate, topicName, message, partition, timestamp)
+                .withProperties(properties)
+                .withHeaders(headers)
+                .withNullKeyRecord(withNullKeyRecord)
+                .build()) {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:67]: no need for this indentation change

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 2]
```diff
@@ -74,46 +77,51 @@ public int sendStringMessagesPlain(long timeoutMs, String topicName, int message
...
-     * @param topicName topic name where messages are send
+     *
+     * @param topicName    topic name where messages are send
      * @param messageCount message count
      * @return sent message count
      */
     public int sendStringMessagesPlain(String topicName, int messageCount) {
         return sendStringMessagesPlain(Duration.ofMinutes(2).toMillis(), topicName, messageCount,
-            List.of(), "\"Hello\" : \"World\"", 0, false);
+                List.of(), "\"Hello\" : \"World\"", 0, null, false);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:87]: no need for this indentation change

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 3]
```diff
@@ -124,11 +132,11 @@ public int sendJsonMessagesPlain(long timeoutMs, String topicName, int messageCo
...
-        try (Producer plainProducer = new Producer.ProducerBuilder(resultPromise, msgCntPredicate, topicName, message, partition)
-            .withProperties(properties)
-            .withHeaders(headers)
-            .withNullKeyRecord(withNullKeyRecord)
-            .build()) {
+        try (Producer plainProducer = new Producer.ProducerBuilder(resultPromise, msgCntPredicate, topicName, message, partition, timestamp)
+                .withProperties(properties)
+                .withHeaders(headers)
+                .withNullKeyRecord(withNullKeyRecord)
+                .build()) {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:139]: no need for this indentation change

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 4]
```diff
@@ -140,73 +148,92 @@ public int sendJsonMessagesPlain(long timeoutMs, String topicName, int messageCo
...
+     * @param messageCount      message count
+     * @param message           specific message to send
+     * @param partition         partition count, how many shards/partitions will topic have
      * @param withNullKeyRecord boolean, which allowing sending messages with NULL key
      * @return sent message count
      */
     public int sendJsonMessagesPlain(String topicName, int messageCount, String message, int partition, boolean withNullKeyRecord) {
         return sendJsonMessagesPlain(Duration.ofMinutes(2).toMillis(), topicName, messageCount, List.of(),
-            message, partition, withNullKeyRecord);
+                message, partition, null,  withNullKeyRecord);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:175]: no need for this indentation change

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 5]
```diff
@@ -140,73 +148,92 @@ public int sendJsonMessagesPlain(long timeoutMs, String topicName, int messageCo
...
-     * @param message specific message to send
-     * @param partition partition count, how many shards/partitions will topic have
+     * @param message      specific message to send
+     * @param partition    partition count, how many shards/partitions will topic have
      * @return sent message count
      */
     public int sendJsonMessagesPlain(String topicName, int messageCount, String message, int partition) {
         return sendJsonMessagesPlain(Duration.ofMinutes(2).toMillis(), topicName, messageCount, List.of(),
-            message, partition, false);
+                message, partition, null, false);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:204]: no need for this indentation change

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 6]
```diff
@@ -140,73 +148,92 @@ public int sendJsonMessagesPlain(long timeoutMs, String topicName, int messageCo
...
+     * @param topicName    topic name where messages are send
      * @param messageCount message count
-     * @param message specific message to send
+     * @param message      specific message to send
      * @return sent message count
      */
     public int sendJsonMessagesPlain(String topicName, int messageCount, String message) {
         return sendJsonMessagesPlain(Duration.ofMinutes(2).toMillis(), topicName, messageCount, List.of(),
-            message, 0, false);
+                message, 0, null, false);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:217]: no need for this indentation change

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java [hunk 7]
```diff
@@ -140,73 +148,92 @@ public int sendJsonMessagesPlain(long timeoutMs, String topicName, int messageCo
...
-     * @param topicName topic name where messages are send
+     *
+     * @param topicName    topic name where messages are send
      * @param messageCount message count
      * @return sent message count
      */
     public int sendJsonMessagesPlain(String topicName, int messageCount) {
         return sendJsonMessagesPlain(Duration.ofMinutes(2).toMillis(), topicName, messageCount, List.of(),
-            "{\"Hello\" : \"World\"}", 0, false);
+                "{\"Hello\" : \"World\"}", 0, null, false);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:229]: no need for this indentation change

**Review by ppatierno — COMMENTED** [review]
I have pushed the doc changes as done last time and left some minor comments.

**Comment by ppatierno** [comment]
@antonio-pedro99 we also need the feature to be listed in the CHANGELOG.

**Review by scholzj — APPROVED** [review]
This seems like something definitely worth having it in the CHANGELOG. LGTM otherwise. Thanks for the PR

**Comment by antonio-pedro99** [comment]
> I have pushed the doc changes as done last time and left some minor comments.

Thanks a lot for this.
I will DM regarding what might be wrong with my setup.

File: src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java (hunk 5 above)
Comment (antonio-pedro99) [reply src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:204]: my IDE's code reformat work :(
I am on it.

Comment (ppatierno) [reply src/test/java/io/strimzi/kafka/bridge/clients/BasicKafkaClient.java:204]: > my IDE's code reformat work :(

Yeah but before pushing you should pay attention on what it did looking at diff, so you can revert the changes.

**Review by ppatierno — APPROVED** [review]
Thanks Antonio. I reworded the CHANGELOG addition and resolved a conflict. Now ready to be merged! LGTM ;-)

**Pull request merged**